### PR TITLE
unify substring for binary&utf8

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #.

# Rationale for this change
 

# What changes are included in this PR?

After arrow-rs have `GenericByteArray`, we can unify `substring` of `binary` & `utf8`.

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-string/src/substring.rs [hunk 1]
```diff
@@ -246,36 +247,59 @@ fn get_start_end_offset(
...
+    let check_char_boundary = {
+        // Safety: a StringArray must contain valid UTF8 data
+        let data_str = unsafe { std::str::from_utf8_unchecked(data) };
+        |offset: T::Offset| {
+            let offset_usize = offset.as_usize();
+            if data_str.is_char_boundary(offset_usize) {
+                Ok(offset)
+            } else {
+                Err(ArrowError::ComputeError(format!(
+                    "The offset {offset_usize} is at an invalid boundary."
```
Comment (alamb) [inline arrow-string/src/substring.rs:273]: I think we should continue to include 'utf-8' in the error message 

```suggestion
                    "The offset {offset_usize} is at an invalid utf-8 boundary."
```

**Review by alamb — COMMENTED** [review]
Thank you @jackwener  -- the idea here is great and reducing code is 💯 

My only concern is about utf8 boundary checking

maybe @tustvold  has some thoughts

File: arrow-string/src/substring.rs [hunk 2]
```diff
@@ -246,36 +247,59 @@ fn get_start_end_offset(
...
+) -> Result<ArrayRef, ArrowError>
+where
+    <T as ByteArrayType>::Native: PartialEq,
+{
     let offsets = array.value_offsets();
     let data = array.value_data();
-    let zero = OffsetSize::zero();
+    let zero = <T::Offset as Zero>::zero();
+
+    // Check if `offset` is at a valid char boundary.
```
[the remark this answers is not in the document]
Comment (jackwener) [reply arrow-string/src/substring.rs:262]: Thanks for @alamb review

File: arrow-string/src/substring.rs [hunk 3]
```diff
@@ -1020,4 +968,17 @@ mod tests {
         let err = substring(&array, 0, Some(5)).unwrap_err().to_string();
         assert!(err.contains("invalid utf-8 boundary"));
     }
+
+    #[test]
+    fn non_utf8_bytes() {
```
Comment (alamb) [inline arrow-string/src/substring.rs:973]: ❤️

**Review by alamb — APPROVED** [review]
Thanks @jackwener  -- looks great to me

**Pull request merged**